### PR TITLE
switch to qtpy

### PR DESCRIPTION
**Pull request by sofroniewn**
# Description
This PR switches to using [qtpy](https://github.com/spyder-ide/qtpy) which provides a standard API for multiple python Qt bindings including both PyQt5 and PySide2 with a single interface. The motivation for this is it gives people, including ourselves the flexibility to either use the GPL licensed PyQt5 or the LGPL license PySide2. When we want to distribute napari as a bundled app the ability to use the LGPL licensed code will enable people to have broader reuse of the code.

Right now though the latest release of vispy only supports `PyQt5`. There is experimental support for `PySide2` on the master of vispy, but I have been unable to get everything working with `PySide2` - I run into the following error after `pip install PySide2`

```
python examples/add_image.py
objc[13156]: Class RunLoopModeTracker is implemented in both /Users/nicholassofroniew/anaconda3/lib/python3.6/site-packages/PySide2/Qt/lib/QtCore.framework/Versions/5/QtCore (0x11e2a2120) and /Users/nicholassofroniew/anaconda3/lib/libQt5Core.5.6.2.dylib (0x12406b8b8). One of the two will be used. Which one is undefined.
WARNING: QObject::moveToThread: Current thread (0x7fa93637bdf0) is not the object's thread (0x7fffc925c3c0).
Cannot move to target thread (0x7fa93637bdf0)

WARNING: You might be loading two sets of Qt binaries into the same process. Check that all plugins are compiled against the right Qt binaries. Export DYLD_PRINT_LIBRARIES=1 and check that only one set of binaries are being loaded.
WARNING: Could not load the Qt platform plugin "cocoa" in "" even though it was found.
WARNING: This application failed to start because no Qt platform plugin could be initialized. Reinstalling the application may fix this problem.

Available platform plugins are: cocoa, minimal, offscreen.

Abort trap: 6
```

Alternatively, if I use conda-forge to install `PySide2` they do not have the latest version and are missing an attribute on the `QtGui` that is present on the pip install version and that vispy requires.

Also how our resources files get built for our QSS stylings might be a little different with `PySide2`

## Type of change

- [x] New feature (non-breaking change which adds functionality)

# How has this been tested?

- [x] running `examples/add_image.py`

## Final checklist:
- [x] My PR is the minimum possible work for the desired functionality
- [ ] I have commented my code, particularly in hard-to-understand areas
- [ ] I have made corresponding changes to the documentation
- [ ] I have added tests that prove my fix is effective or that my feature works

### CODE REVIEW

**Review by kne42 — APPROVED** [review]
lgtm

**Comment by sofroniewn** [comment]
I was able to fix the multiple copies of qt error by running `conda uninstall qt` and am now able to switch between using `pyqt5` and `pyside2` almost without any problems.

I had to turn `app.exec()` into `app.exec_()`, but I don't think that matters. I'm also getting an error about unhashable types that I should be able to fix. The stylesheet stuff seems to be working fine too.

The only problem right now is a vispy one - I get an error `INFO: Program has unset variables: {'texture2D_LUT'}` and the images look terrible, but this is true regardless of `pyqt5` or `pyside2`.

**Comment by sofroniewn** [comment]
I've now fixed the unhashable type error (by adding a `lambda`) so all functionality is preserved with both `pyqt5` and `pyside2` and setting the `QT_API` environment variable appropriately switches between the two options

**Comment by sofroniewn** [comment]
This is ready to be merged - curious if @royerloic, @AhmetCanSolak or @jni have any thoughts on this before it goes in

File: napari/components/_dims/view.py
Comment (jni) [inline napari/components/_dims/view.py]: Erm...

Comment (jni) [reply napari/components/_dims/view.py]: =P

Comment (sofroniewn) [reply napari/components/_dims/view.py]: ;-) fixed

File: napari/components/_window/__init__.py
Comment (jni) [inline napari/components/_window/__init__.py]: What's going on here? Could you add a comment clarifying?

Comment (sofroniewn) [reply napari/components/_window/__init__.py]: hmm yeah, it's maybe not ideal. If you look at the [qtpy](https://github.com/spyder-ide/qtpy) readme you'll see that you can specify which backend to use with the `environ['QT_API']` if multiple backends are present. If multiple are present then it will default to `PyQt5`, if just one is present then it will use that one. The problem is that choice needs to be communicated to vispy. This code will currently cause a conflict if just `PySide2` is present but the environ variable is not set. I'm not quite sure what to do - any advice?

Comment (jni) [reply napari/components/_window/__init__.py]: We could detect it and insert it into os.environ? Is there a way to inspect the backend that qtpy is using?

Comment (jni) [reply napari/components/_window/__init__.py]: Looks like `qtpy3API` might contain the currently-used API?

Comment (jni) [reply napari/components/_window/__init__.py]: qtpy.API, sorry

Comment (jni) [reply napari/components/_window/__init__.py]: Also .API_NAME

Comment (sofroniewn) [reply napari/components/_window/__init__.py]: Great find, thanks! Fixed!

**Review by jni — APPROVED** [review]
Great! LGTM assuming it works. I will have time to test it a bit later, but before merging we should check that it works in an env with just pyside2. Probably we could just add it to the circleCI test matrix?

**Comment by sofroniewn** [comment]
So right now the problem is that the current vispy release doesn't support `PySide2`. The version on master does, and I was able to run napari against that version, though I then ran into other issues (non-qt related) with the version of vispy on master, had to do with textures and openGL stuff.

I'd say that you merge this PR once you've tested it with `PyQt5` and then we investigate backporting the `PySide2` support to vispy or integrating with the new vispy on master and encouraging the vispy team to make a release

**Comment by jni** [comment]
Worked for me! Thanks for handling this @sofroniewn!

**Comment by royerloic** [comment]
This is great @sofroniewn 
I just had a look at qtpy, very nice facade layer for qt.
This is a great design pattern to insulate ourselves from
the idiosyncrasies of particular qt bindings. This has benefits
well beyond licenses. Sometimes bindings 'die', so having a 
facade API protects from that.  

Very cool if vispy can support PySide2.

**Pull request merged**
